### PR TITLE
feat: Set up the Fastapi project

### PR DESCRIPTION
**Pull request by frgfm**
This PR aims at setting up the [fastAPI](https://fastapi.tiangolo.com/) project by:

- setting up a basic fastapi project structure
- adding routes for users, sites, events, devices, media, installations and alerts
- adding CI jobs for lint checking, deps installation, docker container tests

Here is how the documentation renders:
fastapi_docs

Also, once the fastapi server is run, it can be easily accessed using postman for instance:
fastapi_postman

As of now, I implemented the CRUD methods for each route but only added unittests for the "users" route. I'll open up another PR to conduct unittests on the remaining routes.

Any feedback is welcome!

### CODE REVIEW

**Comment by florianriche** [comment]
What led you to try with FastAPI rather than Django ? I've never experienced with FastAPI so I am not very knowledgeable about it.
AFAI see, for our use case, both seem equally easy to use and if we need to do some extensions, Django has more capabilities.

**Comment by frgfm** [comment]
> What led you to try with FastAPI rather than Django ? I've never experienced with FastAPI so I am not very knowledgeable about it.
> AFAI see, for our use case, both seem equally easy to use and if we need to do some extensions, Django has more capabilities.

Hey there,

This PR is mostly to get the choice between the two. The #1 sets up the project as django-based, but the feedback I received from the Data for Good community about choosing django were highly suggestive to try out FastAPI. Performance-wise and dev-wise, so far, it looks like a better option! Besides, the doc building is automatic (OpenAPI & ReDoc both natively supported), and following this [introduction](https://testdriven.io/blog/fastapi-crud/), I was able to implement basic CRUD for the data model, so I'd argue that the learning curve is much nicer than on Django :sweat_smile: 

But this is the whole purpose of having both PRs open, so we can discuss which one would be the best for the project!
@florianriche Which django extensions do you see as essential/relevant to the project? If they are indeed not supported by FastAPI, that could be decisive.

**Comment by florianriche** [comment]
> > What led you to try with FastAPI rather than Django ? I've never experienced with FastAPI so I am not very knowledgeable about it.
> > AFAI see, for our use case, both seem equally easy to use and if we need to do some extensions, Django has more capabilities.
> 
> Hey there,
> 
> This PR is mostly to get the choice between the two. The #1 sets up the project as django-based, but the feedback I received from the Data for Good community about choosing django were highly suggestive to try out FastAPI. Performance-wise and dev-wise, so far, it looks like a better option! Besides, the doc building is automatic (OpenAPI & ReDoc both natively supported), and following this [introduction](https://testdriven.io/blog/fastapi-crud/), I was able to implement basic CRUD for the data model, so I'd argue that the learning curve is much nicer than on Django 
> 
> But this is the whole purpose of having both PRs open, so we can discuss which one would be the best for the project!
> @florianriche Which django extensions do you see as essential/relevant to the project? If they are indeed not supported by FastAPI, that could be decisive.

I had fears regarding SQL DB, admin page, S3 interface. 
But from what I understand from the docs, those are no blockers in FastAPI (although the Django Admin page is nice).

So I have no potential cons to add. The more knowledgeable people from Data4Good seem to think it's a good idea and t's always good to learn something new.

File: src/app/api/alerts/routing.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+@router.post("/", response_model=AlertOut, status_code=201)
+async def create_alert(payload: AlertIn):
+    _id = await crud.post(payload, alerts)
+
+    return {**payload.dict(), "id": _id}
+
+
+@router.get("/{id}/", response_model=AlertOut)
+async def get_alert(id: int = Path(..., gt=0),):
```
Comment (florianriche) [inline src/app/api/alerts/routing.py:19]: I just discovered the "ellipsis" feature thanks to your commit.
What does it do here?

File: src/app/api/alerts/schemas.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+from datetime import datetime
+from pydantic import BaseModel, Field, validator
+from app.db import AlertType
+
+
+class AlertIn(BaseModel):
+    device_id: int = Field(..., gt=0)
```
Comment (florianriche) [inline src/app/api/alerts/schemas.py:7]: Same question as for my previous comment with the "Path". what does the ellipsis do ?

Comment (frgfm) [reply src/app/api/alerts/schemas.py:7]: It's to specify that the field is required! cf. https://pydantic-docs.helpmanual.io/usage/models/#required-fields

File: src/app/api/alerts/routing.py (hunk 1 above)
Comment (frgfm) [reply src/app/api/alerts/routing.py:19]: It specifies that the argument is required

**Comment by frgfm** [comment]
Just updated the project subfolder hierarchy to get something leaner:
```
➜  pyronear-api git:(fastapi-setup) ✗ tree -L 5
├── CONTRIBUTING.md
├── coverage.xml
├── docker-compose.yml
├── LICENSE
├── README.md
└── src
    ├── app
    │   ├── api
    │   │   ├── crud.py
    │   │   ├── routes
    │   │   │   ├── alerts.py
    │   │   │   ├── devices.py
    │   │   │   ├── events.py
    │   │   │   ├── installations.py
    │   │   │   ├── media.py
    │   │   │   ├── ping.py
    │   │   │   ├── sites.py
    │   │   │   └── users.py
    │   │   ├── routing.py
    │   │   └── schemas.py
    │   ├── config.py
    │   ├── db.py
    │   └── main.py
    ├── coverage.xml
    ├── Dockerfile
    ├── requirements.txt
    └── tests
        ├── conftest.py
        ├── __init__.py
        ├── test_ping.py
        └── test_users.py
```

**Comment by frgfm** [comment]
FYI: there is an option to work on an admin page over [there](https://github.com/long2ice/fastapi-admin)

**Comment by florianriche** [comment]
w

**Review by florianriche — DISMISSED** [review]
Looks good to me althought I have a shallow understanding of fastapi.
You may wish to wait for a more knowledgeable reviewer.

**Comment by jeanpasquier75** [comment]
Great job @frgfm I could play with the API thanks to docker-compose really quickly, well done ! FastAPI seems to get more and more partisans so it should not be a problem to find resources/tuto/etc.

Maybe the way it is designed right now is a bit _monolithic_ and **should be okay for MVP**, a more sustainable option could be using _microservices_ like described in [this article](https://dev.to/paurakhsharma/microservice-in-python-using-fastapi-24cc) but it may be not the priority thinking about scalability 🙂

File: README.md [hunk 3]
```diff
@@ -1,2 +1,65 @@
...
+- Python 3.6 (or more recent)
+- [pip](https://pip.pypa.io/en/stable/)
+
+### Installation
+
+You can clone and install the project dependencies as follows:
+
+```bash
+git clone https://github.com/pyronear/pyronear-api.git
+pip install -r pyronear-api/requirements.txt
```
Comment (fe51) [inline README.md:34]: As far as requirements file is in src folder, should it be pip install -r pyronear-api/**_src_**/requirements.txt
```suggestion
 pip install -r pyronear-api/src/requirements.txt
```

**Comment by fe51** [comment]
Nice PR (full of unittest :) ) ! 
I am not very familiar with fastapi yet, but as a start this design suits me (and let's dig a little deeper into the possibility of using microservices)

**Review by fe51 — DISMISSED** [review]

File: README.md (hunk 3 above)
Comment (frgfm) [reply README.md]: Nice catch, just fixed it!

**Review by fe51 — APPROVED** [review]

**Review by florianriche — APPROVED** [review]

**Pull request merged**